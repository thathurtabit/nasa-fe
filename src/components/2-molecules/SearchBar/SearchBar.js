import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import SearchBarStyled, {
  SearchBarInput,
  SearchBarSubmit,
} from './SearchBar.styled';
import SearchIcon from '../../1-atoms/SearchIcon/SearchIcon';
import { SearchPlaceholder } from '../../../utils/constants/constants';
import { setSearch } from '../../../state/actions/setSearch';

const mapStateToProps = state => ({
  searchValue: state.searchValue,
});

const mapDispatchToProps = dispatch => ({
  setSearch: string => dispatch(setSearch(string)),
});

class SearchBar extends Component {
  constructor(props) {
    super(props);

    const { searchValue, setSearch } = props;

    this.setSearch = setSearch;

    this.state = { value: searchValue };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.emitChangeDebounced = debounce(this.emitChange, 300);
  }

  componentWillUnmount() {
    this.emitChangeDebounced.cancel();
  }

  handleChange(event) {
    const { value } = event.target;

    // Update local state to give the user instant feedback...
    this.setState({ value });
    // ... but debounce the result before pushing to redux for performance
    this.emitChangeDebounced(value);
  }

  handleSubmit(event) {
    const { value } = this.state;
    this.setSearch(value);
    event.preventDefault();
  }

  emitChange(value) {
    this.setSearch(value);
  }

  render() {
    const { value } = this.state;
    return (
      <SearchBarStyled role="search" onSubmit={this.handleSubmit}>
        <SearchBarInput
          type="search"
          tabIndex="-1"
          placeholder={SearchPlaceholder}
          aria-label="Search cards"
          value={value}
          onChange={this.handleChange}
        />
        <SearchBarSubmit type="submit" aria-label="Submit">
          <SearchIcon width="20" height="20" title="Search" />
        </SearchBarSubmit>
      </SearchBarStyled>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchBar);

SearchBar.propTypes = {
  setSearch: PropTypes.func.isRequired,
  searchValue: PropTypes.string.isRequired,
};
