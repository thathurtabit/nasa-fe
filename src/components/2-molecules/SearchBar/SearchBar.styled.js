import styled from 'styled-components';

export const SearchBarWrap = styled.section`
  display: flex;
  flex-direction: column;
  position: relative;
`;

export const SearchBarInput = styled.input`
  border: 2px solid ${props => props.theme.input.border};
  border-right: 0;
  font-family: ${props => props.theme.fonts.fontFamily};
  font-size: ${props => props.theme.fonts.baseSize};
  padding: 5px 10px;
  outline: 0;
  text-align: center;
  width: 100%;
  transition: border-color ${props => props.theme.transition.durationFast}
    ease-out;
`;

export const SearchBarSubmit = styled.button`
  align-items: center;
  background: transparent;
  border: 2px solid ${props => props.theme.input.border};
  cursor: pointer;
  display: flex;
  flex-direction: column;
  font-size: ${props => props.theme.fonts.baseSize};
  justify-content: center;
  outline: 0;
  padding: 5px 10px;
  position: relative;
  text-align: center;
  transition: border-color ${props => props.theme.transition.durationFast}
      ease-out,
    background ${props => props.theme.transition.durationFast} ease-out;

  &:hover,
  &:focus {
    background: ${props => props.theme.links.hover};

    svg {
      fill: ${props => props.theme.colors.bodyBg};
    }
  }
`;

const SearchBar = styled.form`
  display: flex;
  margin: 20px auto 40px;
  max-width: 300px;
  position: relative;
  top: -25px;

  &::after {
    border-top: 5px dotted ${props => props.theme.hr.borderColor};
    bottom: -30px;
    content: '';
    left: 50%;
    position: absolute;
    transform: translateX(-50%);
    width: 30px;
  }

  &:hover,
  &:focus {
    input,
    button {
      border-color: ${props => props.theme.input.hover};
    }
  }

  &:focus-within {
    input,
    button {
      border-color: ${props => props.theme.input.focus};
    }
  }
`;

export default SearchBar;
