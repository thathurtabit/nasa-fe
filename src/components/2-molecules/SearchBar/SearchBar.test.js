import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import 'jest-styled-components';
import { shallowWithTheme } from '../../../utils/test_config/testHelpers';
import SearchBar from './SearchBar.styled';

describe('SearchBar test', () => {
  it('should render correctly', () => {
    const element = shallowWithTheme(<SearchBar />);
    expect(shallowToJson(element)).toMatchSnapshot();
  });
});
