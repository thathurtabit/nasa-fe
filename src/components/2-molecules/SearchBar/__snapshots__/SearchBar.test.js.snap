// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar test should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
